### PR TITLE
Add more MIOpen functionality

### PR DESCRIPTION
**Pull request by pxl-th**
- Max/mean pooling.
- Simplify convolutions.
- Softmax.
- Activations.
- Batch Normalization.
- `mark!`/`wait!` on arguments when calling MIOpen functions.

### CODE REVIEW

**Comment by pxl-th** [comment]
Things like dropout *require* passing workspace buffer that was used in forward pass to the backward pass.
Same for batch normalization, which allows passing computed mini-batch mean & inverse variance to the backward pass to speed-up the computation.

From NNlib/Flux we can pass them in rrules and drop during regular calls.

```julia
function rrule()
    y, workspace = MIOpen.dropout(x)
    pullback(dy) = MIOpen.∇dropout(dy, y, x, workspace)
    y, pullback
end
```

File: src/dnn/activations.jl [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    x::ROCArray{T}, desc::ActivationDescriptor,
+) where T <: MIOPENFloat
+    y = similar(x)
+    xdesc, ydesc = TensorDescriptor.((x, y))
+    AMDGPU.wait!(x)
+    miopenActivationForward(
+        handle(), desc.handle, Ref{Float32}(1f0), xdesc.handle, x,
+        Ref{Float32}(0f0), ydesc.handle, y) |> check
+    AMDGPU.mark!(y, C_NULL)
+    AMDGPU.wait!(y)
```
Comment (jpsamaroo) [inline src/dnn/activations.jl:125]: Why are you including this `wait!` call here and below?

File: src/dnn/batchnorm.jl [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    miopenBatchNormalizationForwardInference(
+        handle(), mode, Ref{Float32}(1f0), Ref{Float32}(0f0),
+        xdesc.handle, x, ydesc.handle, y, bndesc.handle,
+        γ, β, μ, ν, ϵ) |> check
+    AMDGPU.mark!(y, C_NULL)
+    AMDGPU.wait!(y)
+    y
+end
+
+function derive_betta_gamma_descriptors(
```
Comment (jpsamaroo) [inline src/dnn/batchnorm.jl:86]:
```suggestion
function derive_beta_gamma_descriptors(
```

Typo?

**Review by jpsamaroo — APPROVED** [review]
LGTM, aside from the double-wait

File: src/dnn/batchnorm.jl (hunk 2 above)
Comment (pxl-th) [reply src/dnn/batchnorm.jl]: Yes, thanks!

File: src/dnn/activations.jl (hunk 1 above)
Comment (pxl-th) [reply src/dnn/activations.jl]: I guess I was overly cautious. Removed them in all places.

Comment (jpsamaroo) [reply src/dnn/activations.jl]: Yeah, if everything using `wait!` and `mark!` correctly, then you only need one pair.
